frostsnap_core/src/sign_task.rs

```diff
@@ -15,19 +15,52 @@ pub enum SignTask {
     BitcoinTransaction(bitcoin_transaction::TransactionTemplate),
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub enum SignTask {
+    Test {
+        message: String,
+    },
+    Nostr {
+        event: Box<crate::nostr::UnsignedEvent>,
+    },
+    BitcoinTransaction {
+        tx_template: bitcoin_transaction::TransactionTemplate,
+        network: bitcoin::Network,
+    },
+}
+
 #[derive(Debug, Clone, PartialEq)]
 /// A sign task bound to a single key. We only support signing tasks with single keys for now.
 pub struct CheckedSignTask {
+    /// The appkey it the task was checked against. Indicates that for example, the Bitcoin
+    /// transaction was signing inputs whose public key was derived from this.
     pub master_appkey: MasterAppkey,
-    pub sign_task: SignTask,
+    pub inner: SignTask,
 }
 
-impl SignTask {
-    pub fn check(self, master_appkey: MasterAppkey) -> Result<CheckedSignTask, SignTaskError> {
-        match &self {
-            SignTask::Plain { .. } | SignTask::Nostr { .. } => {}
-            SignTask::BitcoinTransaction(transaction) => {
-                let non_matching_key = transaction.inputs().iter().find_map(|input| {
+impl WireSignTask {
+    pub fn check(
+        self,
+        master_appkey: MasterAppkey,
+        purpose: KeyPurpose,
+    ) -> Result<CheckedSignTask, SignTaskError> {
+        let variant = match self {
+            WireSignTask::Test { message } => {
+                // We allow any kind of key to sign a test message
+                SignTask::Test { message }
+            }
+            WireSignTask::Nostr { event } => {
+                if !matches!(purpose, KeyPurpose::Nostr) {
+                    return Err(SignTaskError::WrongPurpose);
+                }
+                SignTask::Nostr { event }
+            }
+            WireSignTask::BitcoinTransaction(tx_template) => {
+                let network = match purpose {
```

Comment:
Not related to this PR, but will we eventually want to fix networks on key purposes?
I think there have been some HWW weaknesses in the past regarding a malicious coordinator / social engineering attack by tricking a user into signing "a testnet transaction" that is actually on mainnet.

Reply:
In what way are they not fixed?

Reply:
Ah yes they are fixed in KeyPurpose :+1: